kernel/sleep/src/lib.rs

```diff
@@ -9,128 +9,219 @@
+            return;
+        };
+
+        sleeper.awaken();
+
+        next_wake_time = match sleepers.peek() {
+            Some(next_sleeper) => next_sleeper.wake_time,
+            None => Ticks::MAX,
+        };
     }
-}
 
-/// Remove all tasks that have been delayed but are able to be unblocked now,
-/// the current tick count is provided by the system's interrupt tick count.
-pub fn unblock_sleeping_tasks() {
-    let ticks = TICK_COUNT.load(Ordering::SeqCst);
-    while ticks > NEXT_DELAYED_TASK_UNBLOCK_TIME.load(Ordering::SeqCst) {
-        remove_next_task_from_delayed_tasklist();
-    }
+    // Store the earliest sleeper's wake time for next time.
+    NEXT_WAKE_TIME.store(next_wake_time, Ordering::SeqCst);
 }
 
-/// Blocks the current task by putting it to sleep for `duration` ticks.
-pub fn sleep(duration: usize) {
-    let current_tick_count = TICK_COUNT.load(Ordering::SeqCst);
-    let resume_time = current_tick_count + duration;
 
-    let current_task = current_task().clone();
-    // Add the current task to the delayed tasklist and then block it.
-    add_to_delayed_tasklist(SleepingTaskNode{taskref: current_task.clone(), resume_time});
-    current_task.block();
-    scheduler::yield_now();
+
+/// Blocks asynchronously for `duration` ticks.
+pub async fn sleep(duration: Ticks) {
+    sleep_until(tick_count() + duration).await;
 }
 
-/// Blocks the current task by putting it to sleep until a specific tick count is reached,
-/// given by `resume_time`.
-pub fn sleep_until(resume_time: usize) {
-    let current_tick_count = TICK_COUNT.load(Ordering::SeqCst);
+/// Blocks asynchronously until a specific absolute time.
+/// If the time has already passed there is no delay.
+pub fn sleep_until(wake_time: Ticks) -> impl Future<Output = ()> {
+    let sleeper = Sleeper::new(wake_time);
 
-    if resume_time > current_tick_count {
-        sleep(resume_time - current_tick_count);
+    let mut sleepers = SLEEPERS.lock();
+    sleepers.push(sleeper.clone());
+
+    let next_wake_time = NEXT_WAKE_TIME.load(Ordering::SeqCst);
+    if sleeper.wake_time < next_wake_time {
+        NEXT_WAKE_TIME.store(sleeper.wake_time, Ordering::SeqCst);
     }
+
+    sleeper
 }
 
-/// Blocks the current task for a fixed time `period`, which starts from the given `last_resume_time`.
-pub fn sleep_periodic(last_resume_time: &AtomicUsize, period: usize) {
-    let new_resume_time = last_resume_time.fetch_add(period, Ordering::SeqCst) + period;
-    sleep_until(new_resume_time);
+/// Returns a timer that can be used to sleep for predictable durations correcting for drift.
+/// For example, if you wanted to sleep for precise 100ms intervals
+/// but there could be anywhere from 0-5ms of latency between calls to `sleep`,
+/// this would compensate for the variable delay and sleep up to 5ms less as needed.
+pub fn regular_timer() -> RegularTimer {
+    RegularTimer {
+        last_wake_time: tick_count(),
+    }
+}
```

Comment:
This should probably be an associated `new` method, rather than a separate function.